package/samplers/auto_sampler/sampler.py

```diff
@@ -0,0 +1,220 @@
+            return TPESampler(
+                seed=seed,
+                multivariate=True,
+                warn_independent_sampling=False,
+                constraints_func=self._constraints_func,
+                constant_liar=True,
+            )
+
+        complete_trials = study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,))
+        complete_trials.sort(key=lambda trial: trial.datetime_complete)
+        if len(complete_trials) < self._N_COMPLETE_TRIALS_FOR_CMAES:
+            # Use ``GPSampler`` if search space is numerical and
+            # len(complete_trials) < _N_COMPLETE_TRIALS_FOR_CMAES.
+            if not isinstance(self._sampler, GPSampler):
+                return GPSampler(seed=seed)
+        elif not isinstance(self._sampler, CmaEsSampler):
+            # Use ``CmaEsSampler`` if search space is numerical and
+            # len(complete_trials) > _N_COMPLETE_TRIALS_FOR_CMAES.
+            # Warm start CMA-ES with the first _N_COMPLETE_TRIALS_FOR_CMAES complete trials.
+            warm_start_trials = complete_trials[: self._N_COMPLETE_TRIALS_FOR_CMAES]
+            # NOTE(nabenabe): ``CmaEsSampler`` internally falls back to ``RandomSampler`` for
+            # 1D problems.
+            return CmaEsSampler(
+                seed=seed, source_trials=warm_start_trials, warn_independent_sampling=False
+            )
+
+        return self._sampler  # No update happens to self._sampler.
+
+    def _determine_sampler(
+        self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
+    ) -> BaseSampler:
+        if len(study.directions) == 1:
+            return self._determine_single_objective_sampler(study, trial, search_space)
+        else:
+            return self._determine_multi_objective_sampler(study, trial, search_space)
+
+    def infer_relative_search_space(
+        self, study: Study, trial: FrozenTrial
+    ) -> dict[str, BaseDistribution]:
+        return self._sampler.infer_relative_search_space(study, trial)
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
+    ) -> dict[str, Any]:
+        return self._sampler.sample_relative(study, trial, search_space)
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+        return self._sampler.sample_independent(study, trial, param_name, param_distribution)
+
+    def before_trial(self, study: Study, trial: FrozenTrial) -> None:
+        # NOTE(nabenabe): Sampler must be updated in this method. If, for example, it is updated in
+        # infer_relative_search_space, the sampler for before_trial and that for sample_relative,
+        # after_trial might be different, meaning that the sampling routine could be incompatible.
+        if len(study._get_trials(deepcopy=False, states=(TrialState.COMPLETE,))) != 0:
```

Comment:
How about using `use_cache=True` option?

```suggestion
        if len(study._get_trials(deepcopy=False, states=(TrialState.COMPLETE,), use_cache=True)) != 0:
```